### ISSUE TITLE
Improve the ValidationAdmissionPolicy docs

### ISSUE THREAD

**Issue by samos123** [kubernetes/website#38504]
**This is a Feature Request**

**What would you like to be added**
* End-to-end example: Add example of adding a label environment=test to one of the namespace
* Add explanation to the examples
* Add a note about how to enable the `admissionregistration.k8s.io/v1alpha1` API or link to another doc that talks about runtimeConfig
* remove all the example.com parts since it was mostly confusing
* Link to API docs for the various objects mentioned in the doc
* Move all examples to separate files and ensure they are tested, this would prevent issues like #38495 from occurring again, see comment from @tengqm below 

[comment](https://github.com/kubernetes/website/pull/38497#discussion_r1051312439) from @tengqm:
Issues like this would be easy to detect if we make the manifest an example YAML.
For example:
1. move the YAML snippet into a separate file content/en/examples/policy/validating-admission-ex.yaml and,
2. reference it as {{< codenew file="policy/validating-admission-ex.yaml" >}} in the markdown page,
3. (optionally) amend content/en/examples/examples_test.go so that it can be validated against current and future k8s releases.

**Why is this needed**
Current docs was hard to follow for a new user like myself

**Comment by jimmyraywv** [comment]
So, more documentation about what the admission controller receives from the API would also be helpful. Is it the customary AdmissionReview object?

**Comment by sftim** [comment]
> more documentation about what the admission controller receives from the API would also be helpful. Is it the customary AdmissionReview object?

We typically don't document internal details like that. Maybe that's something we could document within eg https://k8s.dev/docs/

Kubernetes is free to change its internal implementation and end users shouldn't need to know how that's done.

**Comment by sftim** [comment]
/sig api-machinery

Improvements are welcome

This is an alpha feature; we should aim to make the docs better before this graduates to beta.
/priority backlog

**Comment by sftim** [comment]
/language en

If we wanted to, we could add a new tutorial that shows how to do admission-time validation. First with a webhook and then with CEL.

**Comment by jimmyraywv** [comment]
@sftim The issue is that we need to what data is sent in the api server request to the ValidatingAdmissionPolicy.

If you look at the docs for Webhook Mode, you can see that the API server request is documented there: https://kubernetes.io/docs/reference/access-authn-authz/webhook/

The AdmissionReview object is also documented in the Dynamic Admission Control docs: https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#webhook-request-and-response

**Comment by tengqm** [comment]
PR #38522 is adding a reference to the admission config API where AdmissionReview, AdmissionRequest are documented. Is that what you want, @jimmyraywv ?

**Comment by jimmyraywv** [comment]
@tengqm I think that will work. However for [UserInfo v1 authentication.k8s.io](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#userinfo-v1-authentication-k8s-io), for the `groups` []string, are those k8s RBAC groups or are those AuthN IDP groups, like group claims from OIDC? 

What I am driving towards is that will this feature allow us to write policies with CEL that blur the lines between traditional validating admission webhooks, and the AuthZ Webhook Mode?

**Comment by tengqm** [comment]
> @tengqm I think that will work. However for [UserInfo v1 authentication.k8s.io](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#userinfo-v1-authentication-k8s-io), for the `groups` []string, are those k8s RBAC groups or are those AuthN IDP groups, like group claims from OIDC?

AFAICT, k8s doesn't have an abstraction for users or an abstraction for groups. All those information are extracted from the authentication input.

> What I am driving towards is that will this feature allow us to write policies with CEL that blur the lines between traditional validating admission webhooks, and the AuthZ Webhook Mode?

I believe CEL was not designed for authn or authz's purpose. It is a mechanism for you to leverage, to impose constraints that are difficult or impossible to express using OpenAPI schemas. You can mix these admission plugins in any way to meet your requirements.

**Comment by sftim** [comment]
> AFAICT, k8s doesn't have an abstraction for users or an abstraction for groups. All those information are extracted from the authentication input.

You can extend Kubernetes to add those things - this isn't just theoretical, there's a [User](https://docs.openshift.com/container-platform/4.8/rest_api/user_and_group_apis/user-user-openshift-io-v1.html#user-user-openshift-io-v1) API in OpenShift 4.8 for example.

_Core_ Kubernetes doesn't have User or Group APIs. If your container platform does have extension APIs for users and groups, this alpha feature lets you use CEL to validate changes at admission time.

You asked about the (also alpha) SelfSubjectReview API.

https://kubernetes.io/docs/reference/kubernetes-api/authorization-resources/self-subject-review-v1alpha1/#SelfSubjectReviewStatus is filled in after you create a SelfSubjectReview. Here's what you might submit, as YAML, if you wanted to find out who the Kubernetes cluster thinks you are:
```yaml
apiVersion: authentication.k8s.io/v1alpha1
kind: SelfSubjectReview
```
At HTTP level, the body of _response_ might look like:
```json
{
  "apiVersion": "authentication.k8s.io/v1alpha1",
  "kind": "SelfSubjectReview",
  "status": {
    "userInfo": {
      "username": "sftim",
      "uid": "71d99f8d-5b1b-4625-9149-037dc46fdc83",
      "groups": [
        "en-reviewers",
        "en-approvers",
        "system:authenticated"
      ],
      "extra": {
        "emoji": [
          "true"
        ]
      }
    }
  }
}
```

but a ValidatingAdmissionPolicy constrains what you _send_ there, not what you get back.

---

In a couple of days, we'll have a [blog article](https://blog.k8s.io/2022/12/20/validating-admission-policies-alpha/) about this. I'm afraid it's not yet live. Please check back on the 20th of December.

We don't have anything to document here that I can see. Is there a page that needs to be more clear?

**Comment by sftim** [comment]
> The issue is that we need to know what data is sent in the API server request to the ValidatingAdmissionPolicy, so that we can understand what capabilities this solution has.

As a model of its operation, the API server doesn't send any data to itself to implement and enforce a ValidatingAdmissionPolicy. Instead, the ValidatingAdmissionPolicy has access to:
- the entire future object, were admission to succeed
- its parameter data, it any

A ValidatingAdmissionPolicy doesn't have access to any context about the _client_, the _current time_, the _phase of the moon_, etc; if you want to take those into account, you'd need to use a validating admission webhook or some even more custom way to extend Kubernetes.

The docs don't explicitly say this; they explain what is possible and leave you to work the rest out by example. In a tutorial page, if we had one, we could make that point much more explicitly.

**Comment by sftim** [comment]
A new _reference_ page would be good too: we could specify what version of CEL you can use, and the (short) list of available variables.

If we add one, that reference should cover CEL expressions for CRD validation and for ValidatingAdmissionPolicies, in one place. Would that help you @samos123?

**Comment by sftim** [comment]
/reopen

I think this is worth doing.

/sig api-machinery
/remove-lifecycle rotten
/priority backlog

**Comment by cici37** [comment]
/assign @mjlshen 
/reopen
Hi @mjlshen , here is the old issue of suggestions on existing doc style. It would be great if it would squeeze into your doc PR :)

/triage accepted
